### PR TITLE
Partially implement P3697R1 Minor Additions To Standard Library Hardening

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
This PR completes hardening introduced by WG21-P3697R1. Towards #5300.

Hardened preconditions are checked when either `_MSVC_STL_HARDENING_`_`MEOW`_ or `_ITERATOR_DEBUG_LEVEL != 0`. 

Notes:
- The implementation is still partial due to lack of contracts.
- `view_interface` was already hardened in #5274.
- Most checks of `common_iterator` and `counted_iterator` were already done, and this PR only makes the checks enabled with hardening options.
- The hardened preconditions of `basic_stacktrace::operator[]` is checked by default via `vector::operator[]`, so this PR only enable the additional check when the check of `vector::operator[]` is turned off.
- Checks of `shared_ptr::operator[]`, `basic_stacktrace::current`, `common_iterator`'s constructor, `common_iterator::operator==`, `iter_move(counted_iterator)`, and `iter_swap(counted_iterator)` are newly added in this PR.
- A new test entry, `GH_005090_stl_hardening_cxx20_iterators`, is added for C++20 (and potentially later) iterators with `iter_move` and/or `iter_swap` being hidden friends. These iterators are permanently incompatible with `/permissive`.

### CODE REVIEW

**Comment by frederick-vs-ja** [comment]
It seems that we need `strict_20_matrix.lst` for `common_iterator`/`counted_iterator` due to hidden friends `iter_swap` and `iter_move`.

File: stl/inc/yvals.h [hunk 1]
```diff
@@ -178,6 +182,14 @@ _EMIT_STL_ERROR(STL1006, "_CONTAINER_DEBUG_LEVEL has been removed. It was supers
...
 
+#ifndef _MSVC_STL_HARDENING_COMMON_ITERATOR
+#define _MSVC_STL_HARDENING_COMMON_ITERATOR _MSVC_STL_HARDENING
+#endif
+
+#ifndef _MSVC_STL_HARDENING_COUNTED_ITERATOR
+#define _MSVC_STL_HARDENING_COUNTED_ITERATOR _MSVC_STL_HARDENING
+#endif
+
 #ifndef _MSVC_STL_HARDENING_BITSET
```
Comment (StephanTLavavej) [inline stl/inc/yvals.h]: These should be sorted.

File: stl/inc/stacktrace [hunk 2]
```diff
@@ -245,6 +249,10 @@ public:
     }
 
     _NODISCARD const_reference operator[](const size_type _Sx) const noexcept /* strengthened */ {
+        // Bound check is otherwise performed in vector::operator[].
+#if _MSVC_STL_HARDENING_BASIC_STACKTRACE && !_MSVC_STL_HARDENING_VECTOR && _ITERATOR_DEBUG_LEVEL == 0
+        _STL_VERIFY(_Sx < _Frames.size(), "basic_stacktrace subscript out of range");
+#endif // _MSVC_STL_HARDENING_BASIC_STACKTRACE && !_MSVC_STL_HARDENING_VECTOR && _ITERATOR_DEBUG_LEVEL == 0
```
Comment (StephanTLavavej) [inline stl/inc/stacktrace]: This one is interesting because for IDL != 0, it would be better to emit the more specific "basic_stacktrace subscript out of range" message.

I think what we should do here is change the guard to be the normal one (hardening basic stacktrace, or IDL != 0), and then change the access to be unchecked via `vector::data`. Then we get optimal codegen and helpful messages in all cases.

File: tests/std/tests/GH_005090_stl_hardening/test.cpp [hunk 3]
```diff
@@ -19,13 +20,15 @@
 #endif // _HAS_CXX17
 
 #if _HAS_CXX20
+#include <iterator>
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening/test.cpp]: `<iterator>` isn't being tested here - I assume this was a relic from before you had to split out the cxx20_iterators test.

File: tests/std/tests/GH_005090_stl_hardening/test.cpp [hunk 4]
```diff
@@ -587,6 +606,12 @@ void test_mdspan_subscript_span() {
     const span<const int, 2> sp_idx{a_idx};
     (void) md[sp_idx];
 }
+
+// <stacktrace>
+void test_stacktrace_subscript() {
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening/test.cpp]: `basic_stacktrace::current` should also be tested for overflow.

File: tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+    }
+
+    static void reset_throwing() noexcept {
+        throwing_ = false;
+    }
+
+private:
+    static bool check_throwing() {
+        if (throwing_) {
+            throw bad_variant_access{};
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp:181]: Need `<variant>`.

File: tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+        test_common_iterator_assign,
+        test_common_iterator_deref,
+        test_common_iterator_deref_const,
+        test_common_iterator_arrow,
+        test_common_iterator_pre_inc,
+        test_common_iterator_post_inc,
+        test_common_iterator_iter_move,
+        test_common_iterator_iter_swap,
+        test_common_iterator_equality,
+        test_common_iterator_difference,
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp]: equality and difference are defined first, which follows the paper's order.

File: tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+    int n_{};
+
+    int& operator*() { // non-const
+        return n_;
+    }
+
+    int_writer& operator++() {
+        return *this;
+    }
+    int_writer& operator++(int) {
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp]: This doesn't look right - postincrement should return by value.

File: tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+public:
+    using value_type      = int;
+    using difference_type = ptrdiff_t;
+
+    exceptional_iterator() = default;
+    explicit exceptional_iterator(int* p) : ptr_{p} {}
+    exceptional_iterator(const exceptional_iterator& other) : ptr_{other.ptr_} {
+        check_throwing();
+    }
+    exceptional_iterator(exceptional_iterator&) = default;
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp]: This `auto_ptr`-like constructor is extremely confusing. The test breaks if I change it into a defaulted move ctor. As far as I can tell, this `auto_ptr`-like ctor and the defaulted move assign below aren't necessary; if I remove them, making the class copyable-only so everything goes through `check_throwing`, the test passes. I'd like to do that unless there's some reason to have this machinery that I'm missing.

File: tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+    friend ptrdiff_t operator-(int* const se, const exceptional_iterator& i) {
+        return se - i.ptr_;
+    }
+
+    static void reset_throwing() noexcept {
+        throwing_ = false;
+    }
+
+private:
+    static bool check_throwing() {
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp]: The return value of `check_throwing()` is never inspected.

File: tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+        test_counted_iterator_deref,
+        test_counted_iterator_subscript,
+        test_counted_iterator_iter_move,
+        test_counted_iterator_iter_swap,
+        test_counted_iterator_construction,
+        test_counted_iterator_pre_inc,
+        test_counted_iterator_post_inc_input,
+        test_counted_iterator_post_inc_forward,
+        test_counted_iterator_advance,
+        test_counted_iterator_retract,
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp]: Naming nitpick: "advance" is paired with "retreat". ("extend" is paired with "retract".)

File: tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+    exceptional_iterator::reset_throwing();
+    return ret;
+}
+
+// <iterator>
+void test_counted_iterator_deref() {
+    int arr[1]{};
+    auto it = counted_iterator<int*>{arr, 1} + 1;
+    (void) *it;
+}
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp:210]: Should also test the const overload.

File: tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+    jt = it;
+}
+
+void test_common_iterator_deref() {
+    common_iterator<int_writer, default_sentinel_t> it{default_sentinel};
+    (void) *it;
+}
+
+void test_common_iterator_deref_const() {
+    common_iterator<counted_iterator<int*>, default_sentinel_t> it{default_sentinel};
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp]: This needs to be const to test the const overload.

File: tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+}
+
+void test_common_iterator_pre_inc() {
+    common_iterator<counted_iterator<int*>, default_sentinel_t> it{default_sentinel};
+    (void) ++it;
+}
+
+void test_common_iterator_post_inc() {
+    common_iterator<counted_iterator<int*>, default_sentinel_t> it{default_sentinel};
+    (void) ++it;
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp]: This is still testing pre-increment.

File: tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+void test_common_iterator_iter_move() {
+    common_iterator<counted_iterator<int*>, default_sentinel_t> it{default_sentinel};
+    (void) ranges::iter_move(it);
+}
+
+void test_common_iterator_iter_swap() {
+    using CI = counted_iterator<int*>;
+
+    int arr[1]{};
+    common_iterator<CI, default_sentinel_t> it{default_sentinel}, jt{CI{arr, 1}};
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_005090_stl_hardening_cxx20_iterators/test.cpp]: Style: We avoid multi-definitions.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! :heart_eyes_cat: I pushed numerous small changes, the only significant one being to `exceptional_iterator`'s ctor/assign. Please double-check.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
:gem: :gem: :gem:
===
